frontends/PyCDE/src/pycde/module.py

```diff
@@ -204,16 +208,21 @@ def __call__(self, op):
     }
 
     # Build the replacement HWModuleOp in the outer module.
+    module_key = str((op.name, sorted(op_types), sorted(result_types),
```

Comment:
Yes, making a string is gross, but I believe it is the simplest consistent way to do this. Consider that things like `_mlir.ir.Type` are not hashable, and multiple instances of `Types` in Python don't have the same object id like how `Types` can be compared by pointer equality in C++. I considered trying to dig out the underlying `Type` pointer from the `capsule` objects, but that didn't seem possible with the current Python bindings.

Reply:
I'm OK with this.

Reply:
You could always create a `Key` class and override the `__hash__` and `__eq__` methods, but then you'd have to figure out how to hash `mlir.ir.Type`.

Reply:
(I'm not suggesting you do that.)

Reply:
Yep, if `mlir.ir.Type` and `mlir.ir.Attribute` supported `__hash__` I would go that route and have the `Key` hash its members. Will stick to the string key for now.